code/modules/research/xenobiology/crossbreeding/warping.dm

```diff
@@ -0,0 +1,1227 @@
+		var/nlog_type = text2num(log_type)
+		if(nlog_type & LOG_SAY)
+			var/list/reversed = log_source[log_type]  //reverse the list so we get the last sentences instead of the first
+			if(islist(reversed))
+				say_log = reverseRange(reversed.Copy())
+				break
+
+	if(length(say_log) > 10) //we're going to get up to the last 10 sentences spoken by the holo_host
+		recent_speech = say_log.Copy(say_log.len - 11, 0)
+	else
+		for(var/spoken_memory in say_log)
+			if(recent_speech.len >= 10)
+				break
+			recent_speech[spoken_memory] = say_log[spoken_memory]
+
+	if(!length(recent_speech)) //lazy lists don't work here for whatever reason so we set it to null manually if the list is empty.
+		recent_speech = null
+		return
+
+	addtimer(CALLBACK(src, .proc/holo_talk), 100, TIMER_OVERRIDE|TIMER_UNIQUE)
+
+
+///destroys the hologram with the rune
+/obj/effect/warped_rune/ceruleanspace/Destroy()
+	qdel(holotile)
+	holotile = null
+	holo_host = null
+	recent_speech = null
+	return ..()
+
+
+/obj/item/slimecross/warping/pyrite
+	colour = "pyrite"
+	runepath = /obj/effect/warped_rune/pyritespace
+	effect_desc = "draws a rune that will randomly color whatever steps on it"
+
+
+/obj/effect/warped_rune/pyritespace
+	desc = "Who shall we be today? they asked, but not even the canvas would answer."
+	icon_state = "colorune"
+
+
+///colors whatever steps on the rune randomly
+/obj/effect/warped_rune/pyritespace/Crossed(atom/movable/colored)
+	. = ..()
+	colored.color = rgb(rand(0,255),rand(0,255),rand(0,255))
+
+
+/* Will make anyone on the rune do genetic punch damage. May need rebalancing. Not compatible with gloves of the north star*/
+
+
+/obj/item/slimecross/warping/red
+	colour = "red"
+	runepath = /obj/effect/warped_rune/redspace
+	effect_desc = "Draws a rune giving your fists the ability to hurt the very soul of whoever you punch, healing you in the process."
+	drawing_time = 30
+
+
+/obj/effect/warped_rune/redspace
+	desc = "Progress is made through adversity, power is obtained through violence"
+	icon_state = "rage_rune"
+	///fire overlay appearing behind someone when they are on the rune
+	var/mutable_appearance/hate_fire
+
+
+/obj/effect/warped_rune/redspace/Initialize()
+	. = ..()
+	hate_fire = image('icons/mob/OnFire.dmi',"Standing", RAGE_FIRE_LAYER)
+	hate_fire.color = "#C01414" //dark red kind of.
+	hate_fire.layer = RAGE_FIRE_LAYER
+	for(var/mob/living/carbon/human/enraged in rune_turf) //just in case they spawn the rune right under them
+		Crossed(enraged)
+
+
+///boost up the unarmed damage of the person currently on the tile.
+/obj/effect/warped_rune/redspace/Crossed(atom/movable/crossing)
+	. = ..()
+	if(!istype(crossing,/mob/living/carbon/human))
+		return
+
+	var/mob/living/carbon/human/enraged = crossing
+	enraged.add_overlay(hate_fire)
+	RegisterSignal(enraged, COMSIG_HUMAN_EARLY_UNARMED_ATTACK, .proc/genetic_punch)
+	to_chat(enraged, "<span class='warning'>You feel the hate of a thousand souls culminating into your fists!</span>")
+
```

Comment:
Register to `COMSIG_MOVABLE_UNCROSSED` to remove this, instead of checking again in `Uncrossed()`